freestyle/shared/src/main/scala/freestyle/free.scala

```diff
@@ -11,208 +11,163 @@ class free extends StaticAnnotation {
 
 object free {
 
-  def impl(c: whitebox.Context)(annottees: c.Expr[Any]*): c.universe.Tree = {
+  def impl(c: blackbox.Context)(annottees: c.Expr[Any]*): c.universe.Tree = {
     import c.universe._
     import internal.reificationSupport._
 
     def fail(msg: String) = c.abort(c.enclosingPosition, msg)
 
     def gen(): Tree = annottees match {
-      case List(Expr(cls: ClassDef)) => genModule(cls)
+      case List(Expr(cls: ClassDef)) =>
+        val effectTrait @ ClassDef(clsMods, _, _, _) = cls.duplicate
+        if (clsMods.hasFlag(Flag.TRAIT | Flag.ABSTRACT)) {
+          val effectObject = mkEffectObject(effectTrait)
+          q"""
+            $effectTrait
+            $effectObject
+          """
+        } else
+          fail(s"@free requires trait or abstract class")
       case _ =>
         fail(
           s"Invalid @free usage, only traits and abstract classes without companions are supported")
     }
 
-    def genModule(cls: ClassDef) = {
-      val userTrait @ ClassDef(clsMods, clsName, clsParams, clsTemplate) = cls.duplicate
-      if (!clsMods.hasFlag(Flag.TRAIT | Flag.ABSTRACT))
-        fail(s"@free requires trait or abstract class")
-      mkCompanion(clsName.toTermName, clsTemplate.filter {
-        case _: DefDef => true
-        case _         => false
-      }, clsParams, userTrait)
-    }
+    // OP is the name of the Root trait of the Effect ADT
+    lazy val OP = TypeName("T")
```

Comment:
I thought we were going to call the root trait just `Op` literally. I'd like to get rid of the type aliases if we can as I'm having issues finishing the IntelliJ macro support impl because of type aliases. Can we do this in this PR so I can test the plugin with all the new changes and ensure what I'm emitting works after this refactor?https://gitter.im/JetBrains/intellij-scala?at=58bfd7331465c46a56014ac5